File: ui/src/lib/api/prompts.ts
import { request } from "./client";

// OpenRag prompt library (admin-only). Mounted at `/prompts`.
// Shapes verified against openrag/api/schemas/admin/prompt_schemas.py + routers/admin/prompts.py:
//   GET    /prompts/                 list (bare array; optional ?prompt_type=&offset=&limit=)
//   POST   /prompts/                 create → 201
//   GET    /prompts/{id}             get one
//   PATCH  /prompts/{id}             update (name/content/is_default)
//   PUT    /prompts/{id}/default     promote to default for its type
//   DELETE /prompts/{id}             delete → 204
//
// Prompts are *named*; a preset or partition selects one by naming it. There is
// no partition-assignment endpoint — selection lives in the preset/partition
// editors (see presets.tsx / partitions).

// The 8 managed prompt types (mirrors PromptTypeName on the backend).
export type PromptType =
  | "sys_prompt"
  | "query_contextualizer"
  | "chunk_contextualizer"
  | "image_captioning"
  | "hyde"
  | "multi_query"
  | "topic_tagger";

export interface PromptResponse {
  id: string;
  prompt_type: PromptType;
  name: string;
  content: string;
  is_default: boolean;
  created_at: string;
  updated_at: string;
  // Number of partitions that reference this prompt by name (directly for
  // generation prompts, transitively via preset for indexation/retrieval).
  // Only populated by listPrompts(); single-item responses default it to 0.
  used_by: number;
}

export interface CreatePromptRequest {
  prompt_type: PromptType;
  name: string;
  content: string;
  is_default?: boolean;
}

export interface UpdatePromptRequest {
  name?: string;
  content?: string;
  is_default?: boolean;
}

const BASE = "/prompts";
const enc = encodeURIComponent;

/** List library prompts (bare array). Optionally filter by type. */
export function listPrompts(params?: {
  prompt_type?: PromptType;
  offset?: number;
  limit?: number;
}) {
  const search = new URLSearchParams();
  if (params?.prompt_type) search.set("prompt_type", params.prompt_type);
  if (params?.offset !== undefined) search.set("offset", String(params.offset));
  if (params?.limit !== undefined) search.set("limit", String(params.limit));
  const qs = search.toString();
  return request<PromptResponse[]>(`${BASE}/${qs ? `?${qs}` : ""}`);
}

/** Page size used when walking the whole library. The API caps `limit` at 500. */
const PAGE_SIZE = 200;

/** Fetch every library prompt, following offset pagination to the end.
 *
 *  The management page and every prompt picker need the complete library: a
 *  single capped request would silently hide prompts past the cap and report a
 *  partial count as if it were the total, leaving those prompts unmanageable
 *  and unselectable.
 */
export async function listAllPrompts(params?: { prompt_type?: PromptType }): Promise<PromptResponse[]> {
  const all: PromptResponse[] = [];
  for (let offset = 0; ; offset += PAGE_SIZE) {
    const page = await listPrompts({ ...params, offset, limit: PAGE_SIZE });
    all.push(...page);
    // A short page means the end; a full page means there may be more.
    if (page.length < PAGE_SIZE) return all;
  }
}

export function getPrompt(id: string) {
  return request<PromptResponse>(`${BASE}/${enc(id)}`);
}

export function createPrompt(data: CreatePromptRequest) {
  return request<PromptResponse>(`${BASE}/`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function updatePrompt(id: string, data: UpdatePromptRequest) {
  return request<PromptResponse>(`${BASE}/${enc(id)}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

/** Promote a prompt to the default for its type. */
export function setPromptDefault(id: string) {
  return request<PromptResponse>(`${BASE}/${enc(id)}/default`, {
    method: "PUT",
  });
}

export function deletePrompt(id: string) {
  return request<void>(`${BASE}/${enc(id)}`, {
    method: "DELETE",
  });
}

File: openrag/api/routers/admin/prompts.py
"""Admin routes for the DB prompt library.

Transport-only: auth, request validation, and response shaping live here;
persistence and resolution are delegated to ``PromptService`` from the DI
container. Per-partition assignment routes live alongside the other partition
sub-resources in ``partitions.py``.
"""

from api.dependencies.auth import require_admin
from api.schemas.admin.prompt_schemas import (
    CreatePromptRequest,
    PromptResponse,
    PromptTypeName,
    UpdatePromptRequest,
)
from di.providers import get_prompt_service
from fastapi import APIRouter, Depends, Query, Response, status

router = APIRouter(dependencies=[Depends(require_admin)])


@router.post("/", response_model=PromptResponse, status_code=status.HTTP_201_CREATED)
async def create_prompt(
    body: CreatePromptRequest,
    service=Depends(get_prompt_service),
):
    """Add a prompt to the library."""
    return await service.create_prompt(**body.model_dump())


@router.get("/", response_model=list[PromptResponse])
async def list_prompts(
    prompt_type: PromptTypeName | None = None,
    offset: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    service=Depends(get_prompt_service),
):
    """List library prompts (optionally by type), each with an override count."""
    return await service.list_prompts(prompt_type=prompt_type, offset=offset, limit=limit)


@router.get("/{prompt_id}", response_model=PromptResponse)
async def get_prompt(
    prompt_id: str,
    service=Depends(get_prompt_service),
):
    """Return one library prompt."""
    return await service.get_prompt(prompt_id)


@router.patch("/{prompt_id}", response_model=PromptResponse)
async def update_prompt(
    prompt_id: str,
    body: UpdatePromptRequest,
    service=Depends(get_prompt_service),
):
    """Edit a prompt's name/content and/or promote it to default."""
    return await service.update_prompt(prompt_id, **body.model_dump(exclude_unset=True))


@router.put("/{prompt_id}/default", response_model=PromptResponse)
async def set_prompt_default(
    prompt_id: str,
    service=Depends(get_prompt_service),
):
    """Promote a prompt to the default for its type."""
    return await service.set_default(prompt_id)


@router.delete("/{prompt_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_prompt(
    prompt_id: str,
    service=Depends(get_prompt_service),
):
    """Delete a library prompt (rejected if it is the current default)."""
    await service.delete_prompt(prompt_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
